pmd-java/etc/grammar/Java.jjt

```diff
@@ -2415,20 +2452,49 @@ void StatementExpression() :
   PostfixExpression()
 }
 
-void SwitchStatement() :
+void SwitchStatement():
 {}
 {
-  "switch" "(" Expression() ")" "{"
-    ( SwitchLabel() ( BlockStatement() )* )*
-  "}"
+  "switch" "(" Expression() ")" SwitchBlock()
+}
+
+void SwitchBlock() #void :
+{}
+{
+    "{"
+
+        (SwitchLabeledRule())*
+    "}"
 }
 
 void SwitchLabel() :
 {}
 {
-  "case" Expression() ":"
+{ inSwitchLabel = true; }
+(
+  "case" ( ConditionalExpression() #Expression ) ({checkForMultipleCaseLabels();} "," ( ConditionalExpression() #Expression ) )*
 |
-  "default" {jjtThis.setDefault();} ":"
+  "default" {jjtThis.setDefault();}
+)
+{ inSwitchLabel = false; }
+}
+
+// only pushed if it's not a fallthrough, for compatibility
+void SwitchLabeledRule() #SwitchLabeledRule(!inFallthroughSwitchCase):
```

Comment:
This could now be a new switch rule or an old switch label + blockstatements.
Our grammar now allows mixing the two cases within one switch statement, but I think, that's ok.